Sync RST doc

yuyichao opened the pull request:
Also update `genstdlib.jl` to print out the file name in the missing doc warning.

Ref https://github.com/JuliaLang/julia/commit/e08d9797156ceef2ff86032a26be39fa2fa70ca5 https://github.com/JuliaLang/julia/commit/77b2527fd91878d69d69908b60c29d977d2f8e92 https://github.com/JuliaLang/julia/commit/ce21e42b58577670c2ab81da26b53fef51ac8a00 https://github.com/JuliaLang/julia/commit/62027cd50a1f67b5a81be4c1ecf4a829d278e589

Mark as WIP because the doc for the new `qr` method is refering an unexported `qr!` function. Waiting for @jiahao to decide whether it should be exported. If it should be kept unexported, we need to update the rst doc so that it can find `qr!` function in the right module. (or at least export it from `LinAlg` module).

yuyichao committed:
Sync RST doc

Also update genstdlib.jl to print out the file name in the missing doc warning.

yuyichao commented on base/docs/helpdb.jl:
```diff
@@ -328,7 +328,7 @@ Libdl.dlclose
     dlsym_e(handle, sym)
 
-Look up a symbol from a shared library handle, silently return NULL pointer on lookup failure.
+Look up a symbol from a shared library handle, silently return `NULL` pointer on lookup failure.
```
@tkelman This (and the corresponding rst update) is the only one could be backported to 0.4.

yuyichao commented:
Given that the doc refer the function as `qr!` rather than `LinAlg.qr!`, I assume it is meant to be exported. @jiahao could you please confirm that?

I've updated the PR to export those methods and fixed the RST doc to include them. I'll merge this (possibly with the doc in https://github.com/JuliaLang/julia/pull/13841 as well) in a few days if no one want to keep this function unexported.

yuyichao commented on doc/stdlib/linalg.rst:
```diff
@@ -733,8 +833,7 @@ Linear algebra functions in Julia are largely implemented by calling functions f
 
-   .. [issue8859] Issue 8859, "Fix least squares",
-      https://github.com/JuliaLang/julia/pull/8859
+   .. [issue8859] Issue 8859, "Fix least squares", https://github.com/JuliaLang/julia/pull/8859
```
@StefanKarpinski This is updated automatically by the script so I assume it's valid. Just to let you know in case it's not.

andreasnoack commented:
I don't think we should export `qr!` for now. It will probably cause confusion that it doesn't work for matrices so if we want it exprted, then we better make it work for matrices as well.

yuyichao commented:
In that case, should we include it in the document as `LinAlg.qr!`? and should we reference it in the doc of `qr`? I feel like that may also cause confusion for people reading the doc. (Admittedly there's probably more peopling trying things out at REPL than those who carefully going through the online doc ;-p )

andreasnoack commented:
It's a bit weird to document a function without exporting it so maybe we should comment out all mentioning of it until we support `qr!` in general.

yuyichao commented:
It's also a little bit strange to implement a function just to not mention it anywhere.
I could certainly do that if it is what people want though. (Assuming a mutating QR factorization is coming soon this isn't too bad.)

jiahao commented:
I'm not really sure if `qr!` should be exported. I would consider `qr!` an "expert level" function akin to `A_mul_B!`, for people who really want the performance and understand that the input is mutated in place. It's a bit of an oddball function though, since it won't work for matrices as @andreasnoack said - there's usually not a strong motivation for `qr!` as opposed to `qrfact` on a structured orthogonal matrix.

tkelman commented:
Documenting but not exporting is fine. It implies the function is somewhat acceptable for experts to use.

yuyichao commented:
Updated and include the change in #13841.

I removed the export of `qr!`. For `checked_abs` and `qr!`, I included them in the doc as `Base.checked_abs` and `LinAlg.qr!` and fixed other reference to them. I also merged the new and old doc for `abs` since the new one already includes the old one.

tkelman commented on base/linalg/qr.jl:
```diff
@@ -136,15 +136,16 @@ function qr(v::AbstractVector)
 
-Computes the polar decomposition of a vector.
+Computes the polar decomposition of a vector. Instead of returning a new vector
+as `qr(v::AbstractVector)`, this function mutate the input vector `v` in place.
```
mutates

yuyichao replied:
done

yuyichao commented:
CI failed on previous commit for unrelated reason. Rebased and will merge when CI pass.
